client/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -0,0 +1,468 @@
+                                   HttpEntity entity, Header... headers) throws IOException {
+        URI uri = buildUri(endpoint, params);
+        HttpRequestBase request = createHttpRequest(method, uri, entity);
+        setHeaders(request, headers);
+        //we apply a soft margin so that e.g. if a request took 59 seconds and timeout is set to 60 we don't do another attempt
+        long retryTimeoutMillis = Math.round(this.maxRetryTimeoutMillis / (float)100 * 98);
+        IOException lastSeenException = null;
+        long startTime = System.nanoTime();
+        for (HttpHost host : nextHost()) {
+            if (lastSeenException != null) {
+                //in case we are retrying, check whether maxRetryTimeout has been reached, in which case an exception will be thrown
+                long timeElapsedMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTime);
+                long timeout = retryTimeoutMillis - timeElapsedMillis;
+                if (timeout <= 0) {
+                    IOException retryTimeoutException = new IOException(
+                            "request retries exceeded max retry timeout [" + retryTimeoutMillis + "]");
+                    retryTimeoutException.addSuppressed(lastSeenException);
+                    throw retryTimeoutException;
+                }
+                //also reset the request to make it reusable for the next attempt
+                request.reset();
+            }
+
+            CloseableHttpResponse httpResponse;
+            try {
+                httpResponse = client.execute(host, request);
+            } catch(IOException e) {
+                RequestLogger.log(logger, "request failed", request, host, e);
+                onFailure(host);
+                lastSeenException = addSuppressedException(lastSeenException, e);
+                continue;
+            }
+            Response response = new Response(request.getRequestLine(), host, httpResponse);
+            int statusCode = response.getStatusLine().getStatusCode();
+            if (statusCode < 300 || (request.getMethod().equals(HttpHead.METHOD_NAME) && statusCode == 404) ) {
+                RequestLogger.log(logger, "request succeeded", request, host, httpResponse);
+                onSuccess(host);
+                return response;
+            }
+            RequestLogger.log(logger, "request failed", request, host, httpResponse);
+            String responseBody;
+            try {
+                if (response.getEntity() == null) {
+                    responseBody = null;
+                } else {
+                    responseBody = EntityUtils.toString(response.getEntity());
+                }
+            } finally {
+                response.close();
+            }
+            lastSeenException = addSuppressedException(lastSeenException, new ResponseException(response, responseBody));
+            switch(statusCode) {
```

Comment:
The reasoning behind the `500` behavior is that there are no guarantees and no information on which to base a decision. There is no place in elasticsearch (please correct me if I am wrong) that returns 500 willingly, it's just a catch-all for "something went wrong" (I think of it as `RuntimeException`) so we just ignore it.